File: javaScript/24-SwapNodesInPairs.js
"use strict";
console.log('start');
const ListNode = require('./linkedList');

var swapPairs = function(head) {
    let solution = new ListNode();
    let ans = solution;
    let headNext;
    if(head){headNext = head.next;}

    let counter = 0;
    while(headNext){
        solution.next = headNext;
        headNext = headNext.next;
        solution = solution.next;
        solution.next = head;
        solution=solution.next;
        solution.next=null;
        head = headNext;
        if(headNext){headNext = headNext.next};
    }
    solution.next = head;
    return ans.next;
};

let tests = [];
let solutions = [];

tests.push([1,2,3,4,5]);
solutions.push([2,1,4,3,5]);

tests.push([1,2,3,4,5,6]);
solutions.push([2,1,4,3,6,5]);

tests.push([1,2,3,4,5,6,7]);
solutions.push([2,1,4,3,6,5,7]);

tests.push([1,2]);
solutions.push([2,1]);

tests.push([1]);
solutions.push([1]);

tests.push([]);
solutions.push([]);

for (let i=0; i<tests.length;i++){
    let LN = new ListNode();
    LN.createFromArray(tests[i]);
    let result = swapPairs(LN);
    if(result){result = result.toArray();}
    console.log('Test',i,'||',result,solutions[i],result==solutions[i] ? 'PASS':'FAIL');
}

File: 24_Swap_Nodes_in_Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def fromArray(self,array):
        self.val = array[0]
        for i in range(1,len(array)):
            self.next = ListNode()
            self=self.next
            self.val = array[i]
    def toArray(self):
        endArray = []
        head = self
        while head:
            endArray.append(head.val)
            head=head.next
        return endArray
class Solution(object):

    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        if not head or not head.next:
            return head

        nodeNext = head.next
        head.next = self.swapPairs(nodeNext.next)
        nodeNext.next = head
        return nodeNext


tests = []
solutions = []

tests.append([1,2,3,4,5])
solutions.append([2,1,4,3,5])

for i in range(len(tests)):
    list = ListNode()
    list.fromArray(tests[i])
    result = Solution().swapPairs(list)
    print('Test ' + str(i),'Result',result.toArray(),solutions[i])